### PR TITLE
PSR2 WrongOpenercase with colon and bracket is unclear

### LINKED ISSUE #1322: PSR2.ControlStructures.SwitchDeclaration.WrongOpenercase with colon and bracket is unclear
## Describe the bug

The error message is unclear when a switch case begins with `:` and `{`:

> CASE statements must be defined using a colon (PSR2.ControlStructures.SwitchDeclaration.WrongOpenercase)

### Code sample
```php
switch ($type) {
  case 'bool':
  {
    break;
  }
}
```

## Expected behavior

Expect the error message to explain the case cannot include bracket scope after the colon not just that it needs a colon.

## Versions (please complete the following information)

|                                          |                                                                                                                |
|-----------------------------|---------------------------------------------------------------------------- |
| PHP version                      | 8.3                                                                                          |
| PHP_CodeSniffer version | 4.0.1                                                                                  |
| Standard                          | PSR2                                                          |
| Install type                       | composer |

## Please confirm

- [x] I have searched the issue list and am not opening a duplicate issue.
- [x] I have read the [Contribution Guidelines](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/HEAD/.github/CONTRIBUTING.md) and this is not a [support question](https://github.com/PHPCSStandards/PHP_CodeSniffer/discussions).
- [x] I confirm that this bug is a bug in PHP_CodeSniffer and not in one of the external standards.
- [x] I have verified the issue still exists in the `4.x` branch of PHP_CodeSniffer.

### PR DESCRIPTION
**Pull request by bigdevlarry**
Updated SwitchDeclarationSniff to give a clear error when a case uses a braced block after the colon.

# Description

The PR introduces a change that correct updates the SwitchDeclarationSniff to give a clear error when a case uses a braced block after the colon.

## Suggested changelog entry

Improved PSR2 SwitchDeclaration sniff to detect and report case/default statements using a braced block after the colon with (`CaseWithBlockScope`)

## Related issues/external references

Fixes #1322

## Types of changes

- [x] Bug fix _(non-breaking change which fixes an issue)_
- [ ] New feature _(non-breaking change which adds functionality)_
- [ ] Breaking change _(fix or feature that would cause existing functionality to change)_
    - [ ] This change is only breaking for integrators, not for external standards or end-users.
- [ ] Documentation improvement

## PR checklist

- [x] I have checked there is no other PR open for the same change.
- [x] I have read the [Contribution Guidelines](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/HEAD/.github/CONTRIBUTING.md).
- [x] I grant the project the right to include and distribute the code under the BSD-3-Clause license (and I have the right to grant these rights).
- [x] I have added tests to cover my changes.
- [x] I have verified that the code complies with the projects coding standards.
- [ ] \[Required for new sniffs\] I have added XML documentation for the sniff.
- [ ] I have opened a sister-PR in the [documentation repository](https://github.com/PHPCSStandards/PHP_CodeSniffer-documentation) to update the Wiki.

### CODE REVIEW

**Review by jrfnl — COMMENTED** [review]
@bigdevlarry Thanks for opening this PR, but it doesn't correctly address the issue.

1. The PR as-is **_presumes_** this is a block with a colon + curly braces, but doesn't verify it, so there could be a situation where the error message would be shown when there are no curlies, making this a worse situation than before.
2. The error code change is effectively a breaking change as it will break any ignore annotations referencing the original code, as well as ruleset directives referencing the original code.

As things are, I cannot accept this PR.

**Comment by bigdevlarry** [comment]
Hey @jrfnl , thanks for the feedback. I've added a check, and also is there a reasoning for the failing test?

File: src/Standards/PSR2/Sniffs/ControlStructures/SwitchDeclarationSniff.php [hunk 1]
```diff
@@ -197,8 +197,14 @@ public function process(File $phpcsFile, int $stackPtr)
                         }
                     }
                 } else {
-                    // Probably a case/default statement with colon + curly braces.
-                    $phpcsFile->addError($error, $nextCase, 'WrongOpener' . $type);
+                    // Only raise error if colon is followed by a curly brace.
+                    $nextCode = $phpcsFile->findNext(T_WHITESPACE, ($opener + 1), $nextCloser, true);
+                    if ($tokens[$nextCode]['code'] === T_OPEN_CURLY_BRACKET) {
+                        $error = sprintf('%s statements must not use a braced block after the colon', strtoupper($type));
+                        $phpcsFile->addError($error, $nextCase, 'CaseWithBlockScope');
```
Comment (jrfnl) [inline src/Standards/PSR2/Sniffs/ControlStructures/SwitchDeclarationSniff.php]: `addError()` has build-in `sprintf()`-like functionality.

```suggestion
                        $error = '%s statements must not use a braced block after the colon';
                        $phpcsFile->addError($error, $nextCase, 'CaseWithBlockScope', [strtoupper($type)]);
```

**Review by jrfnl — CHANGES_REQUESTED** [review]
> I've added a check

@bigdevlarry Thanks for that, but it looks like you did not (manually) test this change as the code is not working the way you intended it to.
See the code coverage report which shows the code never gets run: https://coveralls.io/builds/77252487/source?filename=src%2FStandards%2FPSR2%2FSniffs%2FControlStructures%2FSwitchDeclarationSniff.php

Also the error code change is still a blocker.

> also is there a reasoning for the failing test?

Rate limit if the broken link check is executed too frequently ;-)

**Comment by bigdevlarry** [comment]
@jrfnl Helpful feedback again; I've now improved the conditional, did some manual test and test coverage should report nicely now :)

File: src/Standards/PSR2/Sniffs/ControlStructures/SwitchDeclarationSniff.php [hunk 2]
```diff
@@ -197,8 +197,17 @@ public function process(File $phpcsFile, int $stackPtr)
                         }
                     }
                 } else {
-                    // Probably a case/default statement with colon + curly braces.
-                    $phpcsFile->addError($error, $nextCase, 'WrongOpener' . $type);
+                    // Only raise error if colon is followed by a curly brace.
+                    $prevToken = $phpcsFile->findPrevious(T_WHITESPACE, ($opener - 1), $nextCase, true);
```
Comment (jrfnl) [inline src/Standards/PSR2/Sniffs/ControlStructures/SwitchDeclarationSniff.php]:
```suggestion
                    $prevToken = $phpcsFile->findPrevious(Tokens::EMPTY_TOKENS, ($opener - 1), null, true);
```

The code as it is now would break on, for example, the following test case (which should be added):
```php
switch ($foo) {
    case 'test': // Trailing comment.
    {
        doSomething();
        break;
    }
}
```

Having said that, the `$opener` code being a `T_OPEN_CURLY_BRACKET` is really the only thing which needs to be checked as the opener will only ever be a `T_OPEN_CURLY_BRACKET` for a "colon - curly" sequence, so line 201 to 205 could be replaced with just a plain:
```php
if ($tokens[$opener]['code'] === T_OPEN_CURLY_BRACKET) {
```

File: src/Standards/PSR2/Sniffs/ControlStructures/SwitchDeclarationSniff.php [hunk 3]
```diff
@@ -197,8 +197,17 @@ public function process(File $phpcsFile, int $stackPtr)
                         }
                     }
                 } else {
-                    // Probably a case/default statement with colon + curly braces.
-                    $phpcsFile->addError($error, $nextCase, 'WrongOpener' . $type);
+                    // Only raise error if colon is followed by a curly brace.
```
Comment (jrfnl) [inline src/Standards/PSR2/Sniffs/ControlStructures/SwitchDeclarationSniff.php]: This comment appears to be in the wrong place as it only applies to the `if`, not to the `else`.

**Review by jrfnl — CHANGES_REQUESTED** [review]
@bigdevlarry Thanks for the update. Not trying to be difficult, but can I please ask you to take another critical look at the code ?

Aside from the inline comments, as I've already previously pointed out, the change in the error code is a breaking change. Please justify why this should have a different error code.

Lastly, unless you can - in all honesty - place a checkmark in the third checkbox in the PR template granting the project the rights to the code, I will not be able to accept the PR.

File: src/Standards/PSR2/Sniffs/ControlStructures/SwitchDeclarationSniff.php (hunk 3 above)
Comment (bigdevlarry) [reply src/Standards/PSR2/Sniffs/ControlStructures/SwitchDeclarationSniff.php]: Removed

File: src/Standards/PSR2/Sniffs/ControlStructures/SwitchDeclarationSniff.php (hunk 2 above)
Comment (bigdevlarry) [reply src/Standards/PSR2/Sniffs/ControlStructures/SwitchDeclarationSniff.php]: Good shout ! I've included the edge case to the test case and reverted to your suggestions. I'm also indifferent about the error code. I agree we should be preventing a breaking change so I've used the `WrongOpener` code.

**Review by jrfnl — APPROVED** [review]
@bigdevlarry Thank you for those changes. All good now. Thank you for contributing.
